crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+                .host()
+                .to_string(),
+            8883,
+        );
+        cloud_config.set_manual_acks(true);
+
+        cloud_config.set_transport(Transport::tls_with_config(tls_config.into()));
+        let topic_prefix = format!("{prefix}/");
+        let (local_client, local_event_loop) = AsyncClient::new(local_config, 10);
+        let (cloud_client, cloud_event_loop) = AsyncClient::new(cloud_config, 10);
+
+        local_client
+            .subscribe(format!("{topic_prefix}#"), QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        for topic in cloud_topics {
+            cloud_client
+                .subscribe(topic.as_ref(), QoS::AtLeastOnce)
+                .await
+                .unwrap();
+        }
+
+        let (tx_pubs_from_cloud, rx_pubs_from_cloud) = mpsc::channel(10);
+        let (tx_pubs_from_local, rx_pubs_from_local) = mpsc::channel(10);
+        tokio::spawn(one_way_bridge(
+            local_event_loop,
+            cloud_client,
+            move |topic| topic.strip_prefix(&topic_prefix).unwrap().into(),
+            tx_pubs_from_local,
+            rx_pubs_from_cloud,
+        ));
+        tokio::spawn(one_way_bridge(
+            cloud_event_loop,
+            local_client,
+            move |topic| format!("{prefix}/{topic}").into(),
+            tx_pubs_from_cloud,
+            rx_pubs_from_local,
+        ));
+
+        Self { signal_sender }
+    }
+
+    pub(crate) fn build_actor(self) -> MqttBridgeActor {
+        MqttBridgeActor {}
+    }
+}
+
+async fn one_way_bridge<F: for<'a> Fn(&'a str) -> Cow<'a, str>>(
+    mut recv_event_loop: EventLoop,
+    target: AsyncClient,
+    transform_topic: F,
+    mut tx_pubs: mpsc::Sender<Publish>,
+    mut rx_pubs: mpsc::Receiver<Publish>,
+) {
+    let mut forward_pkid_to_received_msg = HashMap::new();
+    let mut last_err = None;
+    loop {
+        let notification = match recv_event_loop.poll().await {
+            // TODO notify if this is us recovering from an error
+            Ok(notification) => notification,
+            Err(err) => {
+                let err = err.to_string();
+                if last_err.as_ref() != Some(&err) {
+                    // TODO clarify whether this is cloud/local
+                    error!("MQTT bridge connection error: {err}");
+                    last_err = Some(err);
+                }
+                continue;
+            }
+        };
+        match notification {
+            // Forwarding messages from event loop to target
+            Event::Incoming(Incoming::Publish(publish)) => {
+                target
+                    .publish(
+                        transform_topic(&publish.topic),
+                        publish.qos,
+                        publish.retain,
+                        publish.payload.clone(),
+                    )
+                    .await
+                    .unwrap();
+                tx_pubs.send(publish).await.unwrap();
```

Comment:
Do we also have to do this for QoS 0?

Reply:
> What I'm less sure about is how, if at all, we may wish to protect ourselves against misbehaving MQTT clients. If we aren't receiving acknowledgements, this could result in the map from packet id to original message growing an unbounded amount.

This can be controlled setting a bound to the [number of in flight messages](https://docs.rs/rumqttc/latest/rumqttc/struct.MqttOptions.html#method.set_inflight)